### DISCUSSION TITLE
Feedback for right-to-left language rendering on GitHub

### DISCUSSION THREAD

**Discussion by emmaviolet** [community/community#8115] [Other Feature Feedback, Questions, & Ideas]
Hello! 👋
This is a discussion for gathering user feedback on the right-to-left language support in Markdown, which we [recently released](https://github.blog/changelog/2021-11-16-right-to-left-language-support/).

Please use this discussion to share your experience using the new feature, and any ideas you have on how the experience could be improved.

**Comment by emmaviolet** [comment]
Linking to existing feedback in https://github.com/dear-github/dear-github/issues/147

Notes so far:
- RTL rendering is not currently supported in editors and in titles - this is a known gap but one we'd like to address
- Bullet points with RTL text are not rendered correctly if they have spaces in between the bullets, see https://github.com/dear-github/dear-github/issues/147#issuecomment-975915711 (🙇 @SMSadegh19)

**Reply by jwheare** [reply]
Something to be aware of is URL rendering. You can spoof a domain in the displayed version of the URL by including an RTL override character `[rtlo]` `U+202E` before and within a URL

e.g. (edited for clarity)

`[rtlo]http://example.com/[rtlo]moc.buhtig.www//:sptth`

Is rendered as:

‮http://example.com/‮moc.buhtig.www//:sptth

Which looks like:

https://www.github.com/moc.elpmaxe//:ptth

But is very different, which can be seen if you check the URL tooltip in e.g. chrome:

http://example.com/%E2%80%AEmoc.buhtig.www//:sptth

This can be used as a phishing vector.

Using CSS like the following on autolinked URLs might help mitigate this:

```css
direction: ltr;
unicode-bidi: isolate-override;
```

But URL escaping the [rtlo] before display would probably be wise too.

**Comment by johann-petrak** [comment]
What does "support" actually mean? Is there documentation how to control this? 

What I observe in a MD document that mixes English and arabic words is that if those are mixed, RTL is chosen as soon as one word is Arabic. This is, in my case not what is wanted.

Here is an example

2) أنتِ (anti)

Forms:
* أنتِ is the female singular form (kasra at end)
* أنتَ is the male singular form (fatha at end)
* أنتي sometimes used (incorrectly) for female singular, using "i" instead of kasra
* أنتما dual female or male
* أنتم  plural male
* أنتن plural male


When I check this document in, the WHOLE line for all lines which contain any Arabic word is rendered RTL!! Which is definitely not what I want. 

See https://github.com/johann-petrak/public-sandbox/blob/master/md-ltr-rtl/test1.md

Is there documentation how the direction for a div is determined automatically? Can it be overridden? 

(is this here the only place to provide feedback to github bugs like these? Pity that github provides issue trackers for everyone but apparently not their own software??)

**Reply by johann-petrak** [reply]
So if some random LTR is inserted at the right position, the document gets rendered differently: 
https://github.com/johann-petrak/public-sandbox/blob/master/md-ltr-rtl/test2.md

These heuristics are too simplistic and there needs to be a better way to control this!

**Comment by aghamir** [comment]
Dear @emmaviolet ,
Thanks for your great work.
It seem the task list convert to issue button overlap with the check button in rtl direction

![Screenshot from 2022-05-08 11-45-48](https://user-images.githubusercontent.com/1880925/167286466-b808f0f6-d300-45e1-8397-7726effbdf5b.png)

**Comment by PeterNjeim** [comment]
Block quotes have the vertical bar placed on the wrong side with RTL:
![image](https://user-images.githubusercontent.com/26418974/192591530-0b7b3973-27d9-4e86-bf4f-2ff0e8bff498.png)
It should be on the right side, not the left. This is with `<div dir="rtl" align="right">` as well. Adding an RTL Mark (U+200E) before the `>` doesn't allow the block quote to render.

Another issue in the screenshot above, is the italicized word in the middle of the screenshot. It's slanted the wrong way. It should slant in the direction of reading, so in this case it should slant to the left. Since I'm using Windows, the font used is Segoe UI, and this font doesn't have an italic version of Arabic glyphs, so instead either GitHub or the browser (I'm not sure) is automatically doing a slant, and in this case it's the wrong direction. Instead, it should detect that it's an RTL glyph and slant it in the other direction, if no italic version of the glyph is found.

In the GitHub editor, the `>` symbol is rendered as `<` funnily when doing RTL:
![image](https://user-images.githubusercontent.com/26418974/192592207-e6df5317-b3cd-415a-8be5-3c72cc96d60c.png)
I typed `>` on my keyboard in this instance.

**Comment by PeterNjeim** [comment]
Another issue is that the arrow keys are reversed with RTL text in the GitHub editor. Pressing the left arrow on the keyboard will cause the caret to move to the right. This is probably due to treating the left arrow as a "forward arrow", which isn't the case. The left arrow key should move to the absolute left no matter the text direction, same goes with the right arrow key. This is a bug that occurs in many programs when first implementing RTL, it happened with Notepad++ for example.

**Comment by PeterNjeim** [comment]
Another issue is that the monospace font that is used in the GitHub editor isn't used in the rendered code blocks, instead they are fontless.
![image](https://user-images.githubusercontent.com/26418974/192593286-afc26275-6df0-4101-bbc5-ee8aa35751f6.png)
The Latin text on my Windows PC uses the Consolas font, which I assume is from `ui-monospace`. The Arabic text instead uses Times New Roman, as Windows doesn't come with a monospace Arabic font. The GitHub editor on Windows uses the Courier New font for Arabic monospace, which works but is quite illegible at the same font size as the Consolas Latin text (this is a problem with a majority of Arabic fonts, which give far too much space for diacritics, and try to be "cursive" by making lines diagonal which causes the anti-aliasing to make the glyphs look small). Take a look at this:
![image](https://user-images.githubusercontent.com/26418974/192594627-04725a4b-ad80-4d98-8130-297d5f21c2c9.png)
There's no reason that the hamza (thing above the straight line on the left) should have exclusive access to that area. The straight line (Alif) on the right side should ideally be taller, and whenever a hamza needs to be used, it should "cut into" the Alif's height. This is done by the Times New Roman text in the first screenshot:
![image](https://user-images.githubusercontent.com/26418974/192595258-d94c42c4-56e9-49ab-a4ab-9dfe79abd50c.png)
The straight line on the right side is taller than the straight line on the left side, and instead the hamza "cuts into" it. But it's still not very legible due to the diagonal nature of the font, causing the aliasing to destroy its readability at small font sizes and pixel densities. Here is an example of how the straightness of the Latin text greatly improves its readability:
![image](https://user-images.githubusercontent.com/26418974/192595863-c700900c-9120-4e15-b318-0907f38fbf5d.png)
The Arabic text looks so much smaller simply because of the aliasing.

Of course these font issues are not your fault, however GitHub needs to choose a monospace Arabic font that works in any environment, which ideally means it has proper optical sizing so that small font sizes can adjust and make themselves more legible, and not give exclusive areas to diacritics.

**Comment by ghost** [comment]
> Hello! 👋
> 
> This is a discussion for gathering user feedback on the right-to-left language support in Markdown, which we [recently released](https://github.blog/changelog/2021-11-16-right-to-left-language-support/).
> 
> 
> 
> Please use this discussion to share your experience using the new feature, and any ideas you have on how the experience could be improved.

**Reply by ItaiBEdu** [reply]
**Issue no. 1: A Hebrew paragraph that starts with an English word.**

Correct rendering:
![AHebrewParStartingInEnglish](https://github.com/community/community/assets/143313920/27f696ea-dba3-469a-9906-425fccbe0399)

GitHub's incorrect rendering:

Friends היא תכנית טלוויזיה פופולרית משנות ה- 90.

***

**Issue no. 2: An English phrase ending with a punctuation mark embedded in a Hebrew paragraph.**

Correct rendering:

![EnglishPunctuationMarkInHebrewParagraph](https://github.com/community/community/assets/143313920/f14bb7c2-e656-4ef7-b413-6bd8437f0352)

GitHub's incorrect rendering:

כתוב תכנית שתדפיס: Hello, world!.

***

**Issue no. 3: Syntax-highlighted code blocks in a Hebrew list.**

Correct rendering:
![EnglishCodeblockInHebrewList](https://github.com/community/community/assets/143313920/a8be489d-718e-4114-a5e3-b6870c9a0a52)

GitHub's incorrect rendering:
1. העתיקו את הקוד הבא.
    ```java
    public static void main(String ... args)
    {
        System.out.println("Hello, world!");
    }
    ```

***

**Issue no. 4: A list with *automatically managed* Hebrew numerals.**

Correct rendering:
![ListHebrewNumerals](https://github.com/community/community/assets/143313920/0ac60c84-4510-4720-9ec1-e610d523b417)

How can this be accomplished in GitHub?

**Comment by alidehkhodaei** [comment]
I need this feature. Thank you!
